### PR TITLE
KAFKA-8509; Add downgrade system test

### PR DESCRIPTION
**Pull request by hachikuji**
This patch adds a basic downgrade system test. It verifies that producing and consuming continues to work before and after the downgrade.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by hachikuji** [comment]
Here's a link to a successful run: http://testing.confluent.io/confluent-kafka-branch-builder-system-test-results/?prefix=2019-11-21--001.1574306141--hachikuji--KAFKA-8509--7eee18a/

**Comment by hachikuji** [comment]
I'm changing this test case to use EndToEndTest instead of ProduceConsumeValidateTest. The advantage is that this allows me to assert committed offsets in the consumer, which makes the test case a stronger assertion about compatibility. I will push this change tomorrow.

File: tests/kafkatest/tests/core/downgrade_test.py [hunk 1]
```diff
@@ -21,71 +21,84 @@
...
-    @parametrize(kafka_version=str(LATEST_0_10_1), compression_types=["gzip"])
-    @parametrize(kafka_version=str(LATEST_0_10_0), compression_types=["none"])
-    @parametrize(kafka_version=str(LATEST_0_10_0), compression_types=["lz4"])
-    @parametrize(kafka_version=str(LATEST_0_9), compression_types=["none"], security_protocol="SASL_SSL")
-    @parametrize(kafka_version=str(LATEST_0_9), compression_types=["snappy"])
-    @parametrize(kafka_version=str(LATEST_0_9), compression_types=["lz4"])
-    @parametrize(kafka_version=str(LATEST_0_8_2), compression_types=["none"])
-    @parametrize(kafka_version=str(LATEST_0_8_2), compression_types=["snappy"])
-    def test_upgrade_and_downgrade(self, kafka_version, compression_types, security_protocol="PLAINTEXT"):
+    @parametrize(version=str(LATEST_2_3), compression_types=["none"])
```
Comment (mumrah) [inline tests/kafkatest/tests/core/downgrade_test.py:75]: Could some of these be replaced with `@matrix`?

File: tests/kafkatest/tests/core/downgrade_test.py [hunk 2]
```diff
@@ -21,71 +21,84 @@
...
+    @parametrize(version=str(LATEST_1_0), compression_types=["snappy"])
+    @parametrize(version=str(LATEST_0_11_0), compression_types=["none"])
+    @parametrize(version=str(LATEST_0_11_0), compression_types=["lz4"])
+    @parametrize(version=str(LATEST_0_10_2), compression_types=["none"])
+    @parametrize(version=str(LATEST_0_10_2), compression_types=["lz4"])
+    @parametrize(version=str(LATEST_0_10_1), compression_types=["none"])
+    @parametrize(version=str(LATEST_0_10_1), compression_types=["gzip"])
+    @parametrize(version=str(LATEST_0_10_0), compression_types=["none"])
+    @parametrize(version=str(LATEST_0_10_0), compression_types=["lz4"])
+    @parametrize(version=str(LATEST_0_9), compression_types=["none"], security_protocol="SASL_SSL")
```
Comment (mumrah) [inline tests/kafkatest/tests/core/downgrade_test.py]: Why not PLAINTEXT here as well? Was there something specific to SASL_SSL for this version?

File: tests/kafkatest/tests/produce_consume_validate.py [hunk 3]
```diff
@@ -103,13 +103,14 @@ def stop_producer_and_consumer(self):
         self.producer.stop()
         self.consumer.wait()
 
-    def run_produce_consume_validate(self, core_test_action=None, *args):
-        """Top-level template for simple produce/consume/validate tests."""
+    def run_multistep_produce_consume_validate(self, core_test_actions, *args):
```
Comment (mumrah) [inline tests/kafkatest/tests/produce_consume_validate.py]: Are these changes still needed now that the downgrade test is extending EndToEndTest?

**Review by mumrah — COMMENTED** [review]
Test looks good, few questions inline

File: tests/kafkatest/tests/produce_consume_validate.py (hunk 3 above)
Comment (hachikuji) [reply tests/kafkatest/tests/produce_consume_validate.py]: Thanks, I'll revert these changes.

File: tests/kafkatest/tests/core/downgrade_test.py (hunk 2 above)
Comment (hachikuji) [reply tests/kafkatest/tests/core/downgrade_test.py]: Inherited from the upgrade test. It does seem useful to have some tests which validate the auth flows. For each version, I was considering doing one test with PLAINTEXT/no compression and one with SASL_SSL/some compression. How does that sound?

File: tests/kafkatest/tests/core/downgrade_test.py (hunk 1 above)
Comment (hachikuji) [reply tests/kafkatest/tests/core/downgrade_test.py:75]: Possibly. Let me consider it.

File: tests/kafkatest/services/verifiable_consumer.py [hunk 4]
```diff
@@ -292,8 +295,12 @@ def start_cmd(self, node):
         cmd += self.impl.exec_cmd(node)
         if self.on_record_consumed:
             cmd += " --verbose"
-        cmd += " --reset-policy %s --group-id %s --topic %s --group-instance-id %s --broker-list %s --session-timeout %s --assignment-strategy %s %s" % \
-               (self.reset_policy, self.group_id, self.topic, node.group_instance_id, self.kafka.bootstrap_servers(self.security_config.security_protocol),
+
+        if node.version >= V_2_3_0 and node.group_instance_id:
+            cmd += "--group-instance-id %s" % node.group_instance_id
```
Comment (soondenana) [inline tests/kafkatest/services/verifiable_consumer.py]: nit: may need to add space before `--group-instance-id` as previous if condition adds verbse w/o a trailing space.

**Review by soondenana — COMMENTED** [review]
Thanks for adding these tests, super useful. Left one minor comment.

File: tests/kafkatest/tests/core/downgrade_test.py (hunk 2 above)
Comment (mumrah) [reply tests/kafkatest/tests/core/downgrade_test.py]: If it doesn't add too much to the runtime, I think it would be good to include some more cases like you suggest.

**Comment by hachikuji** [comment]
Finally got a successful run: http://confluent-kafka-branch-builder-system-test-results.s3-us-west-2.amazonaws.com/2019-11-22--001.1574383225--hachikuji--KAFKA-8509--b6c4d3a/report.html.

File: tests/kafkatest/tests/core/downgrade_test.py (hunk 2 above)
Comment (ijuma) [reply tests/kafkatest/tests/core/downgrade_test.py]: Do we want to support all these versions? I'd vote for only testing 1.0 and newer. 1.0 was released roughly 2 years ago.

File: tests/kafkatest/tests/core/downgrade_test.py [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+    @parametrize(version=str(LATEST_0_9), compression_types=["snappy"])
+    @parametrize(version=str(LATEST_0_9), compression_types=["lz4"])
+    def test_upgrade_and_downgrade(self, version, compression_types, security_protocol="PLAINTEXT"):
+        """Test upgrade and downgrade of Kafka cluster from old versions to the current version
+
+        `version` is the Kafka version to upgrade from and dowrade back to
```
Comment (ijuma) [inline tests/kafkatest/tests/core/downgrade_test.py]: Nit: dowrade -> downgrade.

File: tests/kafkatest/tests/core/downgrade_test.py (hunk 2 above)
Comment (ijuma) [reply tests/kafkatest/tests/core/downgrade_test.py]: Given that 1.0 was released 2 years ago, I'd even go with 1.1 as the minimum version.

Comment (hachikuji) [reply tests/kafkatest/tests/core/downgrade_test.py]: Sounds good to me.

**Review by ijuma — APPROVED** [review]
LGTM if we have verified that the system test passes after the latest update.

**Comment by hachikuji** [comment]
Link to successful test run: http://confluent-kafka-branch-builder-system-test-results.s3-us-west-2.amazonaws.com/2019-11-22--001.1574445019--hachikuji--KAFKA-8509--55ac75d/report.html

**Review by mumrah — APPROVED** [review]
